akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/RememberEntitiesStarter.scala

```diff
@@ -47,35 +51,75 @@ private[akka] final class RememberEntityStarter(
     with ActorLogging
     with Timers {
 
-  import RememberEntityStarter.Tick
+  implicit val ec: ExecutionContext = context.dispatcher
+  import RememberEntityStarter._
 
-  private var waitingForAck = ids
-  private var entitiesMoved = Set.empty[ShardRegion.ShardId]
+  private var idsLeftToStart = Set.empty[EntityId]
+  private var waitingForAck = Set.empty[EntityId]
+  private var entitiesMoved = Set.empty[EntityId]
 
-  sendStart(ids)
+  log.debug(
+    "Shard starting [{}] remembered entities using strategy [{}]",
+    ids.size,
+    settings.tuningParameters.entityRecoveryStrategy)
 
-  val tickTask = {
-    val resendInterval = settings.tuningParameters.retryInterval
-    timers.startTimerWithFixedDelay(Tick, Tick, resendInterval)
+  settings.tuningParameters.entityRecoveryStrategy match {
+    case "all" =>
+      idsLeftToStart = Set.empty
+      startBatch(ids)
+    case "constant" =>
+      import settings.tuningParameters
+      idsLeftToStart = ids
+      timers.startTimerWithFixedDelay(
+        "constant",
+        StartBatch(tuningParameters.entityRecoveryConstantRateStrategyNumberOfEntities),
+        tuningParameters.entityRecoveryConstantRateStrategyFrequency)
+      startBatch(tuningParameters.entityRecoveryConstantRateStrategyNumberOfEntities)
   }
+  timers.startTimerWithFixedDelay("retry", ResendUnAcked, settings.tuningParameters.retryInterval)
 
-  def sendStart(ids: Set[ShardRegion.EntityId]): Unit = {
-    // these go through the region rather the directly to the shard
-    // so that shard mapping changes are picked up
-    ids.foreach(id => region ! ShardRegion.StartEntity(id))
+  override def receive: Receive = {
+    case StartBatch(batchSize)                                => startBatch(batchSize)
+    case ShardRegion.StartEntityAck(entityId, ackFromShardId) => onAck(entityId, ackFromShardId)
+    case ResendUnAcked                                        => retryUnacked()
   }
 
-  override def receive: Receive = {
-    case ShardRegion.StartEntityAck(entityId, ackFromShardId) =>
-      waitingForAck -= entityId
-      if (shardId != ackFromShardId) entitiesMoved += entityId
-      if (waitingForAck.isEmpty) {
-        if (entitiesMoved.nonEmpty) shard ! Shard.EntitiesMovedToOtherShard(ids)
-        context.stop(self)
+  private def onAck(entityId: EntityId, ackFromShardId: ShardId): Unit = {
+    idsLeftToStart -= entityId
+    waitingForAck -= entityId
+    if (shardId != ackFromShardId) entitiesMoved += entityId
+    if (waitingForAck.isEmpty && idsLeftToStart.isEmpty) {
+      if (entitiesMoved.nonEmpty) {
+        log.info("Found [{}] entities moved to new shard(s)", entitiesMoved.size)
+        shard ! Shard.EntitiesMovedToOtherShard(entitiesMoved)
       }
+      context.stop(self)
+    }
+  }
+
+  private def startBatch(batchSize: Int): Unit = {
+    log.debug("Starting batch of [{}] remembered entities", batchSize)
```

Comment:
do we want a max outstanding requests?

Reply:
Could make sense but I think we can improve that later if we want it, this is pretty much exactly how it works today.